IsogeomGenerator/refine_isogeom.py

```diff
@@ -0,0 +1,348 @@
+
+    Input:
+    ------
+        fname: string, name of the surface vtk file that will be refined
+        df: float or None, factor for target reduction in the decimation
+            process. Must be between 0 and 1. If set to None, no
+            decimation will be applied. A higher value indicates a
+            greater reduction in the number of surface triangles.
+        sf: float or None, smoothing factor for the smoothing process.
+            Must be between 0 and 1. If set to None, no smoothing will
+            be applied. A higher value indicates more smoothing.
+        surf_type: str, 'interior' or 'exterior', used to indicate if
+            the surface being refined is an interior or exterior surface.
+            Only interior surfaces will have smoothing applied if set.
+
+    Return:
+    -------
+        outfile: string, name of the output file from VTK after refinement
+    """
+    # read in surface w/ vtk as an unstructured grid
+    usg = vtk.vtkUnstructuredGridReader()
+    usg.SetFileName(fname)
+    usg.Update()
+    usgport = usg.GetOutputPort()
+
+    # convert USG to polydata
+    gf = vtk.vtkGeometryFilter()
+    gf.SetInputConnection(usgport)
+    gf.Update()
+    pdport = gf.GetOutputPort()
+
+    # only apply smoothing filter to interior surfaces to avoid smoothing
+    # the corners of the geometry
+    if sf is not None and surf_type == 'interior':
+        smoothfilter = vtk.vtkWindowedSincPolyDataFilter()
+        smoothfilter.SetInputConnection(pdport)
+        smoothfilter.SetNumberOfIterations(20)
+        smoothfilter.BoundarySmoothingOff  # don't apply to boundaries
+        smoothfilter.NonManifoldSmoothingOff()  # don't collapse topology/vols
+        smoothfilter.Update()
+        pdport = smoothfilter.GetOutputPort()
+
+    if df is not None:
+        # setup decimate operator
+        decifilter = vtk.vtkDecimatePro()
+        decifilter.SetInputConnection(pdport)
+        # set decimate options
+        decifilter.SetTargetReduction(df)  # target reduction
+        # preserve topology (splitting or hole elimination not allowed)
+        decifilter.SetPreserveTopology(True)
+        decifilter.SetSplitting(False)  # no mesh splitting allowed
+        # no boundary vertex (eddge/curve) deletion allowed
+        decifilter.SetBoundaryVertexDeletion(False)
+        decifilter.Update()
+        pdport = decifilter.GetOutputPort()
+
+    # convert polydata back to USG
+    appendfilter = vtk.vtkAppendFilter()
+    appendfilter.SetInputConnection(pdport)
+    appendfilter.Update()
+    usg_new = vtk.vtkUnstructuredGrid()
+    usg_new.ShallowCopy(appendfilter.GetOutput())
+
+    outfile = fname.split('.')[0] + '_refined.vtk'
+    writer1 = vtk.vtkGenericDataObjectWriter()
+    writer1.SetFileName(outfile)
+    writer1.SetInputData(usg_new)
+    writer1.Update()
+    writer1.Write()
+
+    return outfile
+
+
+def get_viz_info(mb, surf):
+    """Use MOAB to determine if the triangles of the surface were
```

Comment:
There is no other metadata to transfer. The only metadata that is present on the triangles is `GLOBAL_ID` (which I don't set so that might be something set automatically?) and optionally the data value (ie ww value). When one of the geometries is originally built, the user can request the data to be tagged on each individual triangle for visualization purposes but it isn't necessary at all for tracking (that info is on the surface). So this is checking if the original geometry had that viz data tagged on it and getting the tag information so that it can put the viz data back on the new triangles.

Though originally I was tagging for visualization automatically and getting the data name from the surfaces, not a triangle, which is why I have a long list of tag names that are not the data name. Since really I only need to check on the triangle, this can be simplified slightly.

Reply:
If you have named the tag, can't you just get that tag rather than finding the first tag that isn't something else? It all seems a little convoluted...

Reply:
The tag is named based on the data supplied in the original creation process. So in the refinement process, the name is not automatically known. For example, each energy group on a WW mesh has a different data name: `ww_n_000`, `ww_n_001`, etc. So we have to look to find the tag that is not in this list of standard tags. If such a tag exists on the triangle, then it is assumed to be a data tag that was used for viz like this.